### PR TITLE
Environment_oM: Add Attribute for Plumbing and Fire Protection on Space

### LINKED ISSUE #954: Environment_oM: Adjust Space Object to Include Plumbing and FireProtection Attributes, Adjust Gains to SpaceCriteria
#### Description:
Adjust Gains folder to "SpaceCriteria" it's inclusive of gains (lighting, equipment), as well as acoustics, exhaust, ventilation (remove ventilation folder), profiles, adjust the calls for "gains" to "spacecriteria" in the space object itself

Add plumbing criteria to space object, property on space to be titled "**Plumbing**", the Plumbing object will include:
-**ColdWaterUnits** (double) (purposefully excluding the term Loading or Fixture (units) due to regional differences, flowrate to be calculated via method created later)
-**HotWaterUnits** (double) (purposefully excluding the term Loading or Fixture (units) due to regional differences, flowrate to be calculated via method created later)
-**WasteUnits** (double) (purposefully excluding the term Loading or Fixture (units) due to regional differences, flowrate to be calculated via method created later)
-**VentRequirement** (boolean) (vent is calculated based on the number of wasteunits, to be calculated via method created later)
-**FloorDrainsRequirement** (boolean)
-**HoseBibsRequirment** (boolean) (within spaces that require a hose connection eg Janitors Closet or mechanical room)
-**NaturalGasRate** (double)

Add fire protection criteria to space object, property on space to be titled "**FireProtection**", the object will include:
-**HazardLevel** (enum) (Light, Ordinary, Extra Hazard) (method to follow which calculates the prescriptive number of sprinklers based on the hazard level (eg for an light hazard space, Area/225 SF))
-**CombustibleMaterials** (boolean)
-**SlopedCeiling** (boolean)
-**MaximumCeilingTemperature** (double)
-**SprinklerSystem** (dry, wet, foam, preaction) enum (dry, subject to freezing (non-heated spaces), preaction in art gallery, foam in fuel oil room)

### PR DESCRIPTION
**Pull request by kayleighhoude**
### Issues addressed by this PR

Closes #954 

### Test files

image

### Additional comments
Adjustment from Gains folder to SpaceCriteria will require Dataset refactoring, to follow in subsequent issue.

### CODE REVIEW

File: Environment_oM/Elements/Space.cs [hunk 1]
```diff
@@ -47,28 +46,34 @@ public class Space : BHoMObject, IEnvironmentObject, IElement2D
...
+        public virtual SpaceCriteria.Exhaust Exhaust { get; set; } = new SpaceCriteria.Exhaust();
 
         [Description("Occupant Density is an object that defines the area occupied by each individual in a space (2 m2 per person).")]
         public virtual Occupancy OccupantDensity { get; set; } = new Occupancy();
 
         [Description("Acoustics for an Environments Space define the noise within the space, by noise created by the velocity of the air within the ducts in the space, the Noise Rating, and decibel.")]
-        public virtual Gains.Acoustics Acoustics { get; set; } = new Gains.Acoustics();
+        public virtual SpaceCriteria.Acoustics Acoustics { get; set; } = new SpaceCriteria.Acoustics();
+
+        [Description("The plumbing attributes of a space are indicative of the amount of cold water, hot water and waste required by a space, as well as requirements for vent, floor drains, hose bibs or natural gas.")]
```
Comment (FraserGreenroyd) [inline Environment_oM/Elements/Space.cs:72]: For clarity, it might be prudent to say `waste water` in this descripton.

Comment (FraserGreenroyd) [reply Environment_oM/Elements/Space.cs:72]: Also should `vent` be `vents`?

File: Environment_oM/SpaceCriteria/Enums/HazardType.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+namespace BH.oM.Environment.SpaceCriteria
+{
+    public enum HazardType
```
Comment (FraserGreenroyd) [inline Environment_oM/SpaceCriteria/Enums/HazardType.cs:25]: Can we add a description to this enum? Or on each option to help users understand what they mean?

File: Environment_oM/SpaceCriteria/FireProtection.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        public virtual SprinklerSystemType SprinklerSystemType { get; set; } = SprinklerSystemType.Undefined;
+
+        [Description("The presence of combustible materials in the space would indicate the potential need for increasing the hazard type within the space.")]
+        public virtual bool CombustibleMaterials { get; set; } = false;
+
+        [Description("The presence of a sloped ceiling within the space would indicate the potential need for the area of the ceiling being calculated as if it were flat (i.e. the space would require more sprinklers than by area alone.)")]
+        public virtual bool SlopedCeiling { get; set; } = false;
+
+        [Description("The maximum ceiling temperature indicates the temperature rating, classification and color of the sprinklers within the space.")]
```
Comment (FraserGreenroyd) [inline Environment_oM/SpaceCriteria/FireProtection.cs:50]: `color` should be `colour`.
Can we add a quantity attribute here?

File: Environment_oM/SpaceCriteria/FireProtection.cs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {
+
+        [Description("Hazard Type indicates the combustibility of the materials within the space (Light, Ordinary or Extra). The indication of hazard type also determines the density of sprinklers required within the space.")]
+        public virtual HazardType HazardType { get; set; } = HazardType.Undefined;
+
+        [Description("Sprinkler System Type indicates whether the system will need to accommodate spaces that are subject to freezing (dry system type, commonly used in loading docks and parking garages), spaces that contain materials that when on fire would be spread further by water (foam system type, commonly used in a diesel generator room), spaces that contain precious materials that would be ruined by a false trip of the sprinkler system (preaction, commonly used in art galleries.) ")]
+        public virtual SprinklerSystemType SprinklerSystemType { get; set; } = SprinklerSystemType.Undefined;
+
+        [Description("The presence of combustible materials in the space would indicate the potential need for increasing the hazard type within the space.")]
+        public virtual bool CombustibleMaterials { get; set; } = false;
```
Comment (FraserGreenroyd) [inline Environment_oM/SpaceCriteria/FireProtection.cs:45]: As the description notes the presence of such materials may impact the hazard type, is there a link here that means one property should not be on this object? What would happen if this property is set to true, but hazard type is low or none?

File: Environment_oM/SpaceCriteria/Plumbing.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+using BH.oM.Base;
+using System.ComponentModel;
+
+namespace BH.oM.Environment.SpaceCriteria
+{
+    [Description("The plumbing attributes of a space are indicative of the amount of cold water, hot water and waste required by a space, as well as requirements for vent, floor drains, hose bibs or natural gas")]
+    public class Plumbing : BHoMObject
+    {
+        [Description("The total number of cold water units (aka fixture units or loading units) that should be provided to the space based on the number of plumbing fixtures requiring cold water within the space (e.g. 1 sink = 1.0 units, 2 sinks = 2.0 units, 1 urinal = 10.0 units per IPC2015 Table E103.3).")]
```
Comment (FraserGreenroyd) [inline Environment_oM/SpaceCriteria/Plumbing.cs:37]: Avoid acronyms in descriptions please

File: Environment_oM/SpaceCriteria/Plumbing.cs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        public virtual double ColdWaterUnits { get; set; } = 0.0;
+
+        [Description("The total number of hot water units (aka fixture units or loading units) that should be provided to the space based on the number of plumbing fixtures requiring hot water within the space (e.g. 1 sink = 1.0 units, 2 sinks = 2.0 units, 1 urinal = 0.0 units (no hot water for urinals!) per IPC2015 Table E103.3).")]
+        public virtual double HotWaterUnits { get; set; } = 0.0;
+
+        [Description("The total number of waste/sanitary/drainage units (aka fixture units or loading units) that should be provided to the space based on the number of plumbing fixtures requiring waste/sanitary/drainage connections within the space (e.g. 1 sink = 2.0 units, 2 sinks = 4.0 units, 1 urinal = 4.0 units per IPC2018 Table 709.1).")]
```
Comment (FraserGreenroyd) [inline Environment_oM/SpaceCriteria/Plumbing.cs:43]: Can the units ever be decimal? If not, we should use an `int` here

Comment (kayleighhoude) [reply Environment_oM/SpaceCriteria/Plumbing.cs:43]: Yes, they're occasionally 1.5 etc so a double is more appropriate

File: Environment_oM/Elements/Space.cs (hunk 1 above)
Comment (kayleighhoude) [reply Environment_oM/Elements/Space.cs:72]: Vent in the context of plumbing is just vent (branch vent/stack vent). Waste and waste water are synonyms, sanitary being slightly different in that it comes from a toilet specifically, but I believe lumping all of these together as "waste" would be sufficient as long as it's described properly :)

File: Environment_oM/SpaceCriteria/FireProtection.cs (hunk 4 above)
Comment (kayleighhoude) [reply Environment_oM/SpaceCriteria/FireProtection.cs:45]: Hazard type as a first pass would be set via the designated space type and the presence of combustible materials being a second pass (let's say a gallery that happens to contain something combustible). We can remove the property later if necessary.

**Comment by kayleighhoude** [comment]
Ready for final review @TheresaBrown and @FraserGreenroyd

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
We still need PRs on:

 - BHoM_Engine
 - EnergyPlus_Toolkit
 - TAS_Toolkit

As the changes made her have rippling impacts on those toolkits.

**Review by FraserGreenroyd — APPROVED** [review]
Ok everything is good to go now :smile:

**Pull request merged**